shell/common/shell.cc

```diff
@@ -1223,7 +1223,27 @@ void Shell::LoadDartDeferredLibraryError(intptr_t loading_unit_id,
                                         transient);
 }
 
-void Shell::UpdateAssetManager(std::shared_ptr<AssetManager> asset_manager) {
+void Shell::UpdateAssetResolvers(
+    std::vector<std::unique_ptr<AssetResolver>>& asset_resolvers) {
+  size_t index = 0;
+  auto asset_manager = std::make_shared<AssetManager>();
+  auto old_asset_manager = engine_->GetAssetManager();
+  if (old_asset_manager != nullptr) {
+    for (auto& old_resolver : old_asset_manager->TakeResolvers()) {
+      if (old_resolver->IsUpdatable()) {
+        if (index < asset_resolvers.size()) {
+          // Push the replacement updated resolver in place of the old_resolver.
+          asset_manager->PushBack(std::move(asset_resolvers[index++]));
```

Comment:
Will the `asset_resolvers` argument always contains exactly the same number of updatable resolvers in exactly the same order as the AssetManager's current resolver list?

I'm wondering if it would be simpler to add an API to AssetManager that would replace an AssetResolver of a given type with a new instance.  There would need to be some way to identify which resolver is the APKAssetProvider - perhaps each AssetResolver subclass could return an enum indicating its type.

Reply:
I was hesitant to add too heavy of a solution, but I agree it would be less confusing to explicitly define which resolver type to update.

In most expected circumstances, the number of replacement resolvers would equal the number of updatable resolvers, but i'm not restricting it to be equal here since there is a reasonable course of action for extra resolvers or fewer resolvers. Restricting would necessitate API to reveal the number of updatable resolvers, which isn't necessary for what is trying to be accomplished here.

Reply:
Done.

Reply:
I think it would be sufficient to have UpdateResolversByType accept a single AssetResolver instead of a vector.  The only current use case for UpdateResolversByType is replacing the APK resolver, and the caller expects there to be exactly one of that type in the AssetManager's resolver list.

Reply:
Done